resteasy-core/src/main/java/org/jboss/resteasy/core/providerfactory/ResteasyProviderFactoryImpl.java

```diff
@@ -324,11 +323,8 @@ protected void initialize(ResteasyProviderFactoryImpl parent)
    {
       if (providerClasses == null && parent != null)
          return parent.getProviderClasses();
-      Set<Class<?>> set = new HashSet<Class<?>>();
-      if (parent != null)
-         set.addAll(parent.getProviderClasses());
-      set.addAll(providerClasses);
-      return set;
+
+      return providerClasses;

```

Comment:
The problem I am trying to address is..  When `local` is `true` but `parent` is not `null`, we already create `providerClasses` with parent's [here](https://github.com/resteasy/Resteasy/blob/4.1.1.Final/resteasy-core/src/main/java/org/jboss/resteasy/core/providerfactory/ResteasyProviderFactoryImpl.java#L191).  Then, it gets unnecessary to do a merge with parent here again.

The reason I removed the creation of a new `HashSet` was because there is no possibility of `providerClasses` being `null`.  So, it was always creating the `HashSet`.  But, I had the wrong assumption that if `providerClasses` is not `null`, it always has the parent's, which apparently is incorrect.

For the second scenario you mentioned, should we modify `providerClasses = new CopyOnWriteArraySet<>()` [here](https://github.com/resteasy/Resteasy/blob/4.1.1.Final/resteasy-core/src/main/java/org/jboss/resteasy/core/providerfactory/ResteasyProviderFactoryImpl.java#L162) with `providerClasses = parent == null ? new CopyOnWriteArraySet<>() : new CopyOnWriteArraySet<>(parent.getProviderClasses());` as in the `initialize` method [here](https://github.com/resteasy/Resteasy/blob/4.1.1.Final/resteasy-core/src/main/java/org/jboss/resteasy/core/providerfactory/ResteasyProviderFactoryImpl.java#L191)?

Reply:
I've thought a bit about this, doing what you're proposing would be fine for client-side usage, but likely not that good for the server-side scenario I mentioned. How about keeping track of whether the initialize(..) method has been invoked (a boolean flag member) or not and then using that to decide whether skipping the merge or not in the three get methods here?

Reply:
Sure, I will do that.  But, just for my learning, I have a question.  When `parent` is not `null`, shouldn't we always inherit `providerClasses` from parent to child no matter what?  What is the scenario that we should not do it and why?

Reply:
The scenario is the case here: https://github.com/resteasy/Resteasy/blob/master/resteasy-core/src/main/java/org/jboss/resteasy/core/ResourceMethodInvoker.java#L119 .